### PR TITLE
TLA+: Updating the fastsync spec after model checking

### PR DESCRIPTION
**Pull request by konnov**
This is a new version that:
 * introduces a more refined block model that allows us to check the protocol invariants such as `CorrectBlocksInv`
 * fixes the block execution logic, so the faulty nodes cannot inject corrupted blocks
 * goes through the apalache model checker.

### CODE REVIEW

**Comment by josef-widder** [comment]
Should we add tags?

**Comment by liamsi** [comment]
> This is a new version

This indicates there is a older version that this PR is updating? The diff is indicating that this is mostly new code  (added files only with a tiny exception). I'll add @brapse as a reviewer as he is likely the most familiar with fast-sync here.

I'm happy to review this as well but I think I need someone to walk me through this. It's a really large change and I'm neither deeply familiar with tla nor with fast-sync.

File: docs/spec/fastsync/fastsync.tla [hunk 1]
```diff
@@ -1,4 +1,4 @@
------------------------------ MODULE fastsync -----------------------------
+----------------------------- MODULE fastsync_apalache -----------------------------
```
Comment (konnov) [inline docs/spec/fastsync/fastsync.tla:1]:
```suggestion
----------------------------- MODULE fastsync -----------------------------
```

**Comment by konnov** [comment]
Right. There are plenty of files. The changes to the protocol spec can be found in `fastsync.tla`, whereas the other files are harness for model checking. There a few interesting changes in the protocol that are related to hashes and using a reference chain, which is modeled in `Tinychain.tla`.

File: docs/spec/fastsync/fastsync.tla [hunk 2]
```diff
@@ -148,49 +146,82 @@ VARIABLES
...
 
 
 (********************************** NODE ***********************************)
 
 InitNode ==
-     \E pIds \in SUBSET AllPeerIds \ {{}}:                   \* set of peers node established initial connections with
-        /\ blockPool = [
-                height |-> 1,
+     \E pIds \in SUBSET AllPeerIds:                   \* set of peers node established initial connections with
+        /\ pIds \subseteq CORRECT
```
Comment (josef-widder) [inline docs/spec/fastsync/fastsync.tla:211]: So, no faulty peers initially? Do we need that?

File: docs/spec/fastsync/fastsync.tla [hunk 3]
```diff
@@ -530,40 +574,59 @@ SendStatusResponseMessage(pState) ==
...
 
 SendControlMessage ==
     \/ SendAddPeerMessage
     \/ SendRemovePeerMessage
     \/ SendSyncTimeoutMessage
 
+\* An extremely important property of block hashes (blockId):
+\* If a commit is correct, then the previous block in the block store must be also correct.
+UnforgeableBlockId(height, block) ==
+    block.hashEqRef => block = chain[height]
```
Comment (josef-widder) [inline docs/spec/fastsync/fastsync.tla:595]: We should improve the explanation. It talks about commit mit commit does not show up in formula.

File: docs/spec/fastsync/fastsync.tla [hunk 4]
```diff
@@ -530,40 +574,59 @@ SendStatusResponseMessage(pState) ==
...
+
+\* A faulty peer cannot forge enough of the validators signatures.
```
Comment (josef-widder) [inline docs/spec/fastsync/fastsync.tla:597]:
```suggestion
\* A faulty peer cannot forge enough of the validators signatures.
\* In other words: If a commit contains enough signatures from the validators (in reality 2/3, in the model all), 
\* then the blockID points to the block on the chain, encoded as block.lastCommit.blockIdEqRef being true
```

File: docs/spec/fastsync/fastsync.tla [hunk 5]
```diff
@@ -530,40 +574,59 @@ SendStatusResponseMessage(pState) ==
...
+\* A more precise rule should have checked that the commiters have over 2/3 of the VS's voting power.
+NoFork(height, block) ==
+    (height > 1 /\ block.lastCommit.committers = chain[height - 1].VS)
+        => block.lastCommit.blockIdEqRef
+
+\* Can be block produced by a faulty peer
```
Comment (josef-widder) [inline docs/spec/fastsync/fastsync.tla:603]:
```suggestion
\* Can be block produced by a faulty peer, assuming it cannot generate forks (basic assumption of the protocol)
```

File: docs/spec/fastsync/Tinychain.tla [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    NIL_VS,
+    (* The maximal height, up to which the blockchain may grow *)
+    MAX_HEIGHT
+    
+Heights == 1..MAX_HEIGHT    
+
+\* the set of all possible commits
+Commits == [blockIdEqRef: BOOLEAN, committers: VALIDATOR_SETS]
+
+\* the set of all possible blocks, not necessarily valid ones
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/spec/fastsync/Tinychain.tla:46]: fixed

File: docs/spec/fastsync/fastsync.tla (hunk 2 above)
Comment (konnov) [reply docs/spec/fastsync/fastsync.tla:211]: A faulty peer can be added later, by executing `AddPeer`. However, I agree, we could have faulty peers from the start.

File: docs/spec/fastsync/Tinychain.tla [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+\* 1) the flag of whether the block id equals to the hash of the previous block, and
+\* 2) the set of the validators who have committed the block.
+\* In the implementation, blockId is the hash of the previous block, which cannot be forged.
+\* We abstract block id into whether it equals to the hash of the previous block or not.
+LCT == [blockIdEqRef |-> BOOLEAN, committers |-> VST]
+
+\* Block type.
+\* A block contains its height, validator set, next validator set, and last commit.
+\* Moreover, it contains the flag that tells us whether the block equals to the one
+\* on the reference chain (this is an abstraction of hash).
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/spec/fastsync/Tinychain.tla:25]: Read the isola paper :D

File: docs/spec/fastsync/Tinychain.tla [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+\* the type of validator sets, e.g., STRING
+VST == STRING
+
+\* LastCommit type.
+\* It contains:
+\* 1) the flag of whether the block id equals to the hash of the previous block, and
+\* 2) the set of the validators who have committed the block.
+\* In the implementation, blockId is the hash of the previous block, which cannot be forged.
+\* We abstract block id into whether it equals to the hash of the previous block or not.
+LCT == [blockIdEqRef |-> BOOLEAN, committers |-> VST]
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/spec/fastsync/Tinychain.tla:20]: Read the isola paper :D

**Pull request merged**